### DISCUSSION TITLE
Need help 🙏

### DISCUSSION THREAD

**Discussion by emvaized** [aminomancer/uc.css.js#75] [Q&A] [answered]
I'm trying to write my own first little script, but got stuck with this question — and it seems really hard to find proper documentation on gBrowser and all that stuff these days... 

Could you please tell me, is it possible to add a listener to execute code whenever new "browser" element is about to be added to the browser DOM? I would like to modify some of it's attributes. 
Thanks!

**Comment by aminomancer** [comment]
What attributes are you trying to modify? Some attributes may need to be set before the element connects, others could be set at any point.

**Reply by emvaized** [reply]
`transparent="true"` attribute, in attempt to get fully transparent background for all about: pages

**Reply by emvaized** [reply]
I think it needs to be set before element connects

**Comment by aminomancer** [comment]
Maybe you should override [gBrowser.createBrowser](https://searchfox.org/mozilla-central/rev/241cba7e1f351ddd8a70a8ca05eeb78bca038b75/browser/base/content/tabbrowser.js#2081-2087)?

**Reply by emvaized** [reply]
Yeah, looks like it would be the best way to do it! But is it possible to override this method from user script? I was thinking more about attaching some kind of event listener for browser element creation event

**Comment by aminomancer** [comment]
Yes, it's possible to override it from a user script. Why wouldn't it be?

There is no event that fires precisely when a browser element is created, and even if there was, that wouldn't let you modify it before it connects to the document. It's much better to change how the element is created than to try to hook into its creation from outside.

**Reply by emvaized** [reply]
Okay, got it! Could you please give an example of such gBrowser method override? I think I haven't seen it before in other user scripts

**Comment by aminomancer** [comment, marked as answer]
Check the scripts in my repo, most of them override methods
```js
gBrowser.createBrowser = function (...) {
  ...
}
```

**Reply by emvaized** [reply]
Thanks, it seems I misunderstood this part. Just tried it and got my browser hanged on startup😁 
However, thanks for pointing out! I will try to dig in that direction

**Comment by aminomancer** [comment]
Well you gotta make sure you're overriding the method with a method that actually works. It needs to have all the existing behavior plus the behavior you want to add.

**Reply by emvaized** [reply]
Yeah, makes sense... Thanks for the help! I was using "TabOpen" callback for applying the attribute. But it doesn't work for new `about:newtab` pages and selected tab's `browser` element on Firefox startup (probably because `browser` element gets precached somewhere ommiting the "TabOpen" callback in these cases)

**Comment by aminomancer** [comment]
Yeah TabOpen doesn't seem like a good option for your use case, it represents the creation of a tab element (the actual visual tab) and not the linked browser.

If you wanna get this happening as early as possible, you can probably do something like this.

```js
// ==UserScript==
// @name           Make Browsers Transparent
// @version        1.0.0
// @author         emvaized
// ==/UserScript==

let gBrowser = window.gBrowser ?? window._gBrowser;
if (gBrowser.createBrowser.name !== "customCreateBrowser") {
  eval(
    `gBrowser.createBrowser = function customCreateBrowser` +
      gBrowser.createBrowser
        .toSource()
        .replace(/^createBrowser/, "")
        .replace(/(\s*)(type: \"content\",)/, `$1$2$1transparent: "true",`)
  );
}
```

However, I'm not sure this gets called for the first browser tab at startup. The first tab in a window is special in a lot of ways. You'll have to do some trial and error probably.

The later parts in the readme on my repo has a lot of useful links on learning how to do these kinds of mods. The best advice is to get familiar with searching for stuff on searchfox :P

**Reply by emvaized** [reply]
Thanks, I'll check this!

**Comment by emvaized** [comment]
I just tested the script you provided, and it works just [perfect](https://i.imgur.com/ElPpLxM.png)! Even for the first browser tab at startup. Huge thanks👍
I only had to rename all "gBrowser" to just "gBrow", because somehow got an error in console on trying to reassign global variable. 

However, now a new task arose in my head — it seems sidebar's `browser` element is not affected by this patch, as it seems to be using [another method](https://searchfox.org/mozilla-central/source/browser/base/content/webext-panels.js#20) for `browser` creation. I guess, this one couldn't be overriden in the same way, as it's not part of globally exposed object like `gBrowser`?

**Reply by emvaized** [reply]
By reading the source code I suggest that maybe it could be overriden by referencing it by SidebarUI class, like it's seen [here](https://searchfox.org/mozilla-central/source/browser/base/content/browser-sidebar.js#495) for `loadPanel` method? 
However, I just tried to override it the same way, but I get `SidebarUI.browser.contentWindow.getBrowser is undefined` on browser startup. Maybe this method is initialized some time later, not immediately on startup?

**Reply by aminomancer** [reply]
gBrowser is only a global object for the chrome window. Other browser elements are totally unrelated to it, so if you want to override their creation, you'll need another bespoke solution for those. There are lots of things that [create browsers](https://searchfox.org/mozilla-central/search?q=createxulelement%28%22browser&path=&case=false&regexp=false).

If it's actually true that you need to set `transparent` before the element connects, then overriding these things individually is the best you can do. For the sidebar you'd probably need to set transparent on both [this](https://searchfox.org/mozilla-central/rev/2bd7d039738b1aab07c47549652be5ff9d06329d/browser/base/content/webext-panels.js#34) and [this](https://searchfox.org/mozilla-central/rev/2bd7d039738b1aab07c47549652be5ff9d06329d/browser/base/content/browser-box.inc.xhtml#17). The latter is easy, since there's only ever one of those per chrome window. But the former is no easy feat, as that script is loaded inside the latter. It's possible to do this but it's gonna involve a lot of research, and I don't have that kind of time to spare.

If it's not true that you need to set `transparent` before the element connects, then you could skip all that and just override the [custom element definition](https://searchfox.org/mozilla-central/source/toolkit/content/widgets/browser-custom-element.js) for `<browser>`. That would be ideal. In order to override a custom element, you need to actually override the file itself. Which you can't do from a script. You need to use [utils/chrome.manifest](https://github.com/aminomancer/uc.css.js/blob/master/utils/chrome.manifest) to change the file that the URL `chrome://global/content/elements/browser-custom-element.js` points to. You have to redirect it to your own version of the file, copying it [from searchfox](https://searchfox.org/mozilla-central/source/toolkit/content/widgets/browser-custom-element.js) and changing the [construct method](https://searchfox.org/mozilla-central/rev/2bd7d039738b1aab07c47549652be5ff9d06329d/toolkit/content/widgets/browser-custom-element.js#1009) to add `this.setAttribute("transparent", "true");`

There are examples of that kind of custom element override [in this repo](https://github.com/aminomancer/uc.css.js/tree/master/resources/script-override). You can see all those script-override files are referenced in my chrome.manifest file. So you need to have the files in your resources folder, and you need to add the actual URL override to chrome.manifest. Then, when Firefox tries to access `chrome://global/content/elements/browser-custom-element.js`, instead of finding the original file bundled with Firefox, it will find the custom modified file in your resources folder instead. Then you'll have overridden all browser elements across the entire app.

**Reply by aminomancer** [reply]
I misread your comment by the way. Thought you said gBrowser not getBrowser. Yes, you're half right. It's not that the method is initialized after startup, it's that `SidebarUI.browser.contentWindow` doesn't exist when the app starts. `SidebarUI` does, but `SidebarUI.browser.contentWindow` is a reference to, well, the content window. There is no content when the sidebar is not showing, so there is no window. And every time the window shows, it reloads webext-panels.xhtml, which reloads webext-panels.js from disk. So you wouldn't be able to just override it once. You would need to somehow listen for the sidebar opening and override that method every time it opens, which probably wouldn't work anyway because that method is getting used as soon as the sidebar opens. So chances are it's running a few milliseconds before you can override it, and even if not, that's still a race condition, so it's probably unreliable.

The thing you're trying to do just isn't really an easy project. Might be something to come back to after you're more familiar with the code base and have a good sense of how everything fits together. But it's possible the chrome.manifest approach works for what you're trying to do, and that would certainly be very easy compared to trying to override every method that creates a browser with JS. So I would give that a try at least

**Reply by emvaized** [reply]
Thanks for your explanation. Yeah, definetely the sidebar with "webext-panels-browser" id should also be modified for my purpose. 

Guess I'm out of luck then//. After some research I only found this [SidebarShown](https://searchfox.org/mozilla-central/source/browser/base/content/browser-sidebar.js#419) event — which is good since it seems to be fired [before](https://searchfox.org/mozilla-central/source/browser/base/content/browser-sidebar.js#551) the "webext-panels-browser" sidebar gets loaded and inserted into it. But trying to overwrite the `SidebarUI.browser.contentWindow.getBrowser` method there results in errors that [promiseEvent](https://searchfox.org/mozilla-central/source/browser/base/content/webext-panels.js#18) and [ExtensionUtils](https://searchfox.org/mozilla-central/source/browser/base/content/webext-panels.js#14) are undefined...

**Reply by emvaized** [reply]
I managed to do it! By setting my userscript to be run only in `chrome://browser/content/webext-panels.xhtml` documents via `@include` option, I managed to overwrite `window.getBrowser` function in the same way you showed previously.

**Reply by aminomancer** [reply]
Nice! Out of curiosity, did you ever try the manifest method I described earlier?

**Reply by emvaized** [reply]
Nah, I didn't, as it seemed too much for me to overwrite the whole file just for adding that one attribute... Also I had worries that on Firefox updates potential changes will not be included in the overwritten file, and I will have to do it manually each time. I decided to consider it as a plan B to try later, if I won't find a way to overwrite the function in the runtime — which feels much more simple and sustainable to me.

**Reply by emvaized** [reply]
Guess I will publish the script later and create some corresponding repo, as there are many folks out there who request such changes occasionally – in most cases, to get the fully transparent background for TreeStyleTabs or Sidebery sidebars.
